compiler/src/main/java/com/squareup/anvil/compiler/codegen/BindingModuleGenerator.kt

```diff
@@ -0,0 +1,236 @@
+          }
+      val contributedBindingsDependencies = classScanner.findContributedClasses(
+          module,
+          packageName = HINT_BINDING_PACKAGE_PREFIX,
+          annotation = contributesBindingFqName
+      )
+
+      val bindingMethods = (contributedBindingsThisModule + contributedBindingsDependencies)
+          .filter {
+            val annotation = it.annotationOrNull(contributesBindingFqName)
+            annotation != null && scope == annotation.scope(module).fqNameSafe
+          }
+          .map { contributedClass ->
+            val annotation = contributedClass.annotation(contributesBindingFqName)
+            val boundType = annotation.boundType(module)
+
+            checkExtendsBoundType(type = contributedClass, boundType = boundType)
+
+            val concreteType = contributedClass.fqNameSafe
+            val methodName = concreteType
+                .asString()
+                .split(".")
+                .joinToString(separator = "", prefix = "bind") { it.capitalize(US) }
+
+            val paramName = concreteType.shortName()
+                .asString()
+                .decapitalize(US)
+
+            """
+              |
+              |  @$daggerBindsFqName abstract fun $methodName(
+              |    $paramName: $concreteType
+              |  ): ${boundType.fqNameSafe}
+            """.trimMargin()
+          }
+
+      daggerModuleFiles.forEach { (file, psiClass) ->
+        file.writeText(
+            daggerModuleContent(
+                scope = scope.asString(),
+                psiClass = psiClass,
+                bindingMethods = bindingMethods
+            )
+        )
+      }
+    }
+  }
+
+  private fun findContributedBindingClasses(projectFiles: Collection<KtFile>) {
+    contributedBindingClasses += projectFiles
+        .filter {
+          it.packageFqName.asString()
+              .startsWith(HINT_BINDING_PACKAGE_PREFIX)
+        }
+        .flatMap {
+          it.findChildrenByClass(KtProperty::class.java)
+              .toList()
+        }
+        .mapNotNull { ktProperty ->
+          ((ktProperty.initializer as? KtClassLiteralExpression)
+              ?.firstChild as? KtDotQualifiedExpression)
+              ?.text
+              ?.let { FqName(it) }
+        }
+  }
+
+  private fun checkExtendsBoundType(
+    type: ClassDescriptor,
+    boundType: ClassDescriptor
+  ) {
+    val boundFqName = boundType.fqNameSafe
+    val hasSuperType = type.getAllSuperClassifiers()
+        .any { it.fqNameSafe == boundFqName }
+
+    if (!hasSuperType) {
+      throw AnvilCompilationException(
+          classDescriptor = type,
+          message = "${type.fqNameSafe} contributes a binding for $boundFqName, but doesn't " +
+              "extend this type."
+      )
+    }
+  }
+
+  private fun generateClassName(psiClass: KtClassOrObject): String =
+    psiClass.parentsWithSelf
+        .filterIsInstance<KtClassOrObject>()
+        .toList()
+        .reversed()
```

Comment:
Why are you reversing the list?

Reply:
`com.abc.MyClass.InnerClass.InnerInnerClass`. Calling `parents`walks the structure from right to left. But I want classes from left to right. I want to generate `MyClassInnerClassInnerInnerClassAnvilModule`. Without reversing I'd generate `InnerInnerClassInnerClassMyClassAnvilModule`.